In [2]:
#Importing Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression

#Importing the dataset
data = pd.read_csv("diabetes.csv")
data.head(10)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   
5            5      116             74              0        0  25.6   
6            3       78             50             32       88  31.0   
7           10      115              0              0        0  35.3   
8            2      197             70             45      543  30.5   
9            8      125             96              0        0   0.0   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  
5                     0.201   30        0  
6                     0.248   26        1  
7                     0.134   29        0  
8                     0.158   53        1  
9                     0.232   54        1

## Identify the IV's and DV

In [3]:
X = data.iloc[:,[0,1,2,3,4,5,6,7]].values
y = data.iloc[:,[8]].values 

## Training and Test Samples

In [6]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.2)

print ("Training data:",X_train.shape, y_train.shape)
print ("Test data:",X_test.shape, y_test.shape)

Training data: (614, 8) (614, 1)
Test data: (154, 8) (154, 1)


In [7]:
X_train

array([[  0.   , 167.   ,   0.   , ...,  32.3  ,   0.839,  30.   ],
       [  7.   , 181.   ,  84.   , ...,  35.9  ,   0.586,  51.   ],
       [  4.   , 134.   ,  72.   , ...,  23.8  ,   0.277,  60.   ],
       ...,
       [  4.   , 129.   ,  60.   , ...,  27.5  ,   0.527,  31.   ],
       [  0.   , 117.   ,   0.   , ...,  33.8  ,   0.932,  44.   ],
       [  2.   ,  91.   ,  62.   , ...,  27.3  ,   0.525,  22.   ]])

## Applying Scalers to the training and test sample sets

In [135]:
# importing scaling modules
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler

### 1.  MinMaxScaler

In [136]:
# Creating an object of the MinMaxScalar function
scaler = MinMaxScaler()

# Fitting and transforming scaler fucntion on the training dataset
X_train = scaler.fit_transform((X_train))

minmax_df_new = pd.DataFrame(X_train, columns = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 
                                                   'BMI', 'DiabetesPedigreeFunction', 'Age'])
minmax_df_new

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0       0.411765  0.474747       0.581818       0.252525  0.106183  0.581152   
1       0.000000  0.525253       0.690909       0.000000  0.000000  0.321117   
2       0.176471  0.560606       0.527273       0.313131  0.059140  0.514834   
3       0.117647  0.631313       0.545455       0.202020  0.188172  0.589878   
4       0.058824  0.686869       0.672727       0.505051  0.274194  0.652705   
..           ...       ...            ...            ...       ...       ...   
609     0.235294  0.863636       0.654545       0.000000  0.000000  0.760908   
610     0.529412  0.732323       0.727273       0.464646  0.174731  0.661431   
611     0.176471  0.308081       0.745455       0.282828  0.000000  0.600349   
612     0.235294  0.454545       0.000000       0.000000  0.000000  0.488656   
613     0.529412  0.535354       0.472727       0.000000  0.000000  0.544503   

     DiabetesPedigreeFunction       Age  
0                    0.293203  0.333333  
1                    0.223900  0.100000  
2                    0.156375  0.016667  
3                    0.004442  0.166667  
4                    0.142603  0.050000  
..                        ...       ...  
609                  0.178143  0.083333  
610                  0.248334  0.316667  
611                  0.073301  0.416667  
612                  0.236339  0.166667  
613                  0.134163  0.350000  

[614 rows x 8 columns]

In [137]:
# Transforming the test dataset using the training dataset fit 
X_test = scaler.transform(X_test)
X_test_minmax_new = pd.DataFrame(X_test, columns = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 
                                                   'BMI', 'DiabetesPedigreeFunction', 'Age'])
X_test_minmax_new

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0       0.294118  0.727273       0.745455       0.262626  0.383065  0.558464   
1       0.058824  0.439394       0.545455       0.373737  0.100806  0.649215   
2       0.117647  0.500000       0.472727       0.151515  0.126344  0.429319   
3       0.411765  0.808081       0.490909       0.323232  0.235215  0.532286   
4       0.294118  0.444444       0.709091       0.303030  0.000000  0.481675   
..           ...       ...            ...            ...       ...       ...   
149     0.176471  0.560606       0.818182       0.121212  0.104839  0.495637   
150     0.529412  0.863636       1.000000       0.242424  0.322581  0.792321   
151     0.117647  0.565657       0.600000       0.222222  0.000000  0.436300   
152     0.235294  0.459596       0.636364       0.323232  0.118280  0.577661   
153     0.411765  0.505051       0.000000       0.000000  0.000000  0.523560   

     DiabetesPedigreeFunction       Age  
0                    0.166148  0.616667  
1                    0.191470  0.016667  
2                    0.248334  0.000000  
3                    0.226566  0.300000  
4                    0.079964  0.266667  
..                        ...       ...  
149                  0.185251  0.133333  
150                  0.285651  0.550000  
151                  0.101733  0.050000  
152                  0.163483  0.016667  
153                  0.180364  0.183333  

[154 rows x 8 columns]

### 2. StandardScaler

In [138]:
scaler2 = StandardScaler()
X_train = scaler2.fit_transform((X_train))
standard_df_new = pd.DataFrame(X_train, columns = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age'])
standard_df_new

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0       0.977590 -0.840928      -0.252788       0.300025 -0.006834  0.192760   
1      -1.159583 -0.527202       0.365997      -1.259537 -0.694539 -1.733700   
2      -0.243652 -0.307594      -0.562181       0.674320 -0.311513 -0.298552   
3      -0.548962  0.131622      -0.459050      -0.011887  0.524179  0.257406   
4      -0.854273  0.476720       0.262866       1.859588  1.081308  0.722859   
..           ...       ...            ...            ...       ...       ...   
609     0.061659  1.574760       0.159735      -1.259537 -0.694539  1.524473   
610     1.588211  0.759073       0.572259       1.610058  0.437128  0.787505   
611    -0.243652 -1.876223       0.675390       0.487173 -0.694539  0.334981   
612     0.061659 -0.966418      -3.552978      -1.259537 -0.694539 -0.492491   
613     1.588211 -0.464457      -0.871574      -1.259537 -0.694539 -0.078755   

     DiabetesPedigreeFunction       Age  
0                    0.854849  0.646313  
1                    0.361165 -0.537011  
2                   -0.119859 -0.959627  
3                   -1.202165 -0.198918  
4                   -0.217963 -0.790580  
..                        ...       ...  
609                  0.035208 -0.621534  
610                  0.535220  0.561790  
611                 -0.711646  1.068929  
612                  0.449775 -0.198918  
613                 -0.278091  0.730836  

[614 rows x 8 columns]

In [139]:
X_test = scaler2.transform(X_test)
X_test_std_new = pd.DataFrame(X_test, columns = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 
                                                   'BMI', 'DiabetesPedigreeFunction', 'Age'])
X_test_std_new

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0       0.366969  0.727701       0.675390       0.362408  1.786424  0.024679   
1      -0.854273 -1.060536      -0.459050       1.048615 -0.041654  0.697001   
2      -0.548962 -0.684065      -0.871574      -0.323800  0.123743 -0.932086   
3       0.977590  1.229662      -0.768443       0.736703  0.828859 -0.169260   
4       0.366969 -1.029164       0.469128       0.611938 -0.694539 -0.544208   
..           ...       ...            ...            ...       ...       ...   
149    -0.243652 -0.307594       1.087914      -0.510947 -0.015539 -0.440774   
150     1.588211  1.574760       2.119223       0.237643  1.394693  1.757200   
151    -0.548962 -0.276222      -0.149657       0.112878 -0.694539 -0.880369   
152     0.061659 -0.935046       0.056604       0.736703  0.071513  0.166901   
153     0.977590 -0.652693      -3.552978      -1.259537 -0.694539 -0.233906   

     DiabetesPedigreeFunction       Age  
0                   -0.050237  2.083206  
1                    0.130147 -0.959627  
2                    0.535220 -1.044150  
3                    0.380153  0.477267  
4                   -0.664177  0.308220  
..                        ...       ...  
149                  0.085842 -0.367965  
150                  0.801050  1.745114  
151                 -0.509110 -0.790580  
152                 -0.069225 -0.959627  
153                  0.051031 -0.114395  

[154 rows x 8 columns]

## Creating the Multiple Linear Regression Model

In [140]:
X_train = X_train.reshape(-1, 8)
y_train = y_train.reshape(-1, 1) 

# create an empty linear regression model like below and give it a good variable name
diabetes_model = LinearRegression()

# to create the model, we use fit(X,y)
model = diabetes_model.fit(X_train,y_train)

print('Intercept: \n', model.intercept_)
print('Coefficients: \n', model.coef_)

Intercept: 
 [0.34364821]
Coefficients: 
 [[ 0.06581657  0.18220107 -0.04605465 -0.00020818 -0.03730136  0.09979911
   0.0494126   0.02867235]]


### Modelled linear equation for dependent variable prediction

* **outcome** = 0.34 + 0.066(Pregnancies) + 0.18(Glucose) - 0.046(BloodPressure) - 0.0002(SkinThickness) - 0.037(Insulin) + 0.1(BMI) + 0.049(DPF) + 0.029(Age)

### Predictions for the Test set

In [141]:
predictions = diabetes_model.predict(X_test)

In [142]:
predictions[0:5]

array([[0.46228343],
       [0.16514514],
       [0.1219587 ],
       [0.65193175],
       [0.10616774]])

In [143]:
predictions[predictions < 0.5] = 0
predictions[predictions >= 0.5] = 1

In [144]:
predictions[0:5]

array([[0.],
       [0.],
       [0.],
       [1.],
       [0.]])

## Accuracy Check of the Predictive Model

In [145]:
from sklearn.metrics import accuracy_score

In [146]:
acc = accuracy_score(y_test, predictions)
acc

0.8116883116883117

The model is achieving an predictive accuracy of 81%.